### PR TITLE
Opt constraints tests

### LINKED ISSUE #46: Documentation for Optimizer
There are a couple things.

First, the gain parameters all have virtually identical documentation, which makes them difficult to differentiate, especially when one does not have access to the reference paper cited.

Second, there is a hard-coded input space for the SolutionExport of an Optimizer-based step, using the input "trajID", but this isn't mentioned in the user manual.  I couldn't find an example in the user guide either.  Some explanation of the necessary mechanics would be quite helpful.

Third, the node `<parameter>` is incorrectly listed as <Parameter> in the documentation (the example is correct).

Fourth, in the `<Functions>` block description, it indicates absence of Functions and the `<upperBound>` nodes results in an unconstrained problem; however, the `<upperBound>` nodes are indicated as required for each variable.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement? ***improvement***
- [x] 2. Is it tagged with a priority: critical, normal or minor? ***normal***
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements? ***n/a***
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users. ***not a defect***
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.) ***yes***

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #46.

##### What are the significant changes in functionality due to this change request?
 - Fixes optimization constraints, and adds an analytic constraint test, to the repository
 - Adds an 11-page user guide entry on using the optimizers
 - Fixes several ambiguities and incorrect statements in the user manual regarding optimization 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/Optimizers/Optimizer.py [hunk 1]
```diff
@@ -436,7 +436,7 @@ def initialize(self,externalSeeding=None,solutionExport=None):
     for trajInd in self.optTraj:
       for varname in self.getOptVars():
         varK[varname] = self.optVarsInit['initial'][varname][trajInd]
-      satisfied, _ = self.checkConstraint(self.normalizeData(varK))
+      satisfied, _ = self.checkConstraint(varK)
```
Comment (alfoa) [inline framework/Optimizers/Optimizer.py]: Thanks

File: tests/framework/Optimizers/gold/Stochastic/opt_soln.csv [hunk 2]
```diff
@@ -1,2 +1,6 @@
 trajID,y,x,ans,varsUpdate
+<<<<<<< HEAD
+0.0,1.74245729318,-1.72247273512,25.355339271,84.0
+=======
 0.0,1.23350487406,-1.63980124837,19.5238896591,22.0
+>>>>>>> devel
```
Comment (alfoa) [inline tests/framework/Optimizers/gold/Stochastic/opt_soln.csv]: conflict did not get solved here.

File: tests/framework/user_guide/optimizing/multilevel.xml [hunk 3]
```diff
@@ -0,0 +1,73 @@
+<?xml version="1.0" ?>
+<Simulation>
```
Comment (alfoa) [inline tests/framework/user_guide/optimizing/multilevel.xml]: ```<TestInfo>``` missing

File: tests/framework/user_guide/optimizing/multiple_trajectories.xml [hunk 4]
```diff
@@ -0,0 +1,70 @@
+<?xml version="1.0" ?>
+<Simulation>
+  <RunInfo>
+    <WorkingDir>MultipleTraj</WorkingDir>
+    <Sequence>optimize,print</Sequence>
+  </RunInfo>
```
Comment (alfoa) [inline tests/framework/user_guide/optimizing/multiple_trajectories.xml]: ```<TestInfo>``` missing

File: tests/framework/user_guide/optimizing/simple.xml [hunk 5]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" ?>
+<Simulation>
+  <RunInfo>
+    <WorkingDir>Simple</WorkingDir>
+    <Sequence>optimize,print</Sequence>
+  </RunInfo>
```
Comment (alfoa) [inline tests/framework/user_guide/optimizing/simple.xml]: ```<TestInfo>``` missing

File: tests/framework/user_guide/optimizing/step_size.xml [hunk 6]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" ?>
+<Simulation>
+  <RunInfo>
+    <WorkingDir>StepSize</WorkingDir>
+    <Sequence>optimize,print</Sequence>
+  </RunInfo>
```
Comment (alfoa) [inline tests/framework/user_guide/optimizing/step_size.xml]: as above

File: tests/framework/user_guide/optimizing/summary.xml [hunk 7]
```diff
@@ -0,0 +1,90 @@
+<?xml version="1.0" ?>
+<Simulation>
+  <RunInfo>
+    <WorkingDir>Summary</WorkingDir>
+    <Sequence>optimize,getOptPoint,print</Sequence>
+  </RunInfo>
```
Comment (alfoa) [inline tests/framework/user_guide/optimizing/summary.xml]: as above. Sorry for multiple messages

Comment (PaulTalbot-INL) [inline tests/framework/user_guide/optimizing/summary.xml]: No problem!  Sorry I forgot this syntax.

**Comment by alfoa** [comment]
The reason of the failure for ```.\tests\framework\user_guide\ForwardSamplingStrategies.Grid``` is because the test uses 8 processors and sometimes the one history finishes before the other determining that the diff fails.

@PaulTalbot-INL  even if this does not relate to this MR, can you change the test to use a batchsize of 1 or change the CSV we compare to (using UnorderedCSV)?

**Comment by PaulTalbot-INL** [comment]
Ok, sure.

**Comment by alfoa** [comment]
Ok.

Comments have been addressed...

Once the tests pass, this PR can be accepted

**Comment by PaulTalbot-INL** [comment]
@bobkinl With the two merges today, we're set for another push to Hybrid, if you have the time this week.

**Comment by bobkinl** [comment]
I'm on it.

**Pull request merged**